migrate and update `callPackage` tutorial (#935)
NixOS/nix.dev: source/tutorials/callpackage.md

---
date: 2022-09-08
authors:
  - Norbert Melzer
  - Valentin Gagarin
  - Matthias Meschede
myst:
  html_meta:
    "keywords": "tutorial, callPackage, override, package, customise, parameters, nix, nixpkgs"
---

(callpackage-tutorial)=
# Package parameters and overrides with `callPackage`

Nix ships with a a special-purpose programming language for creating packages and configurations: the Nix language.
It is used to build the Nix package collection, known as {term}`Nixpkgs`.

Being purely functional, the Nix language allows declaring custom functions to abstract over common patterns.
One of the most prominent patterns in Nixpkgs is parametrisation of package recipes.

## Overview

Nixpkgs is sizeable software project on it's own, with coding conventions and idioms that have emerged over the years.

> It has [established a convention](https://github.com/NixOS/nixpkgs/pull/9869) of composing parameterised packages with automatic settings through a function named [`callPackage`](https://github.com/NixOS/nixpkgs/commit/fd268b4852d39c18e604c584dd49a611dc795a9b).

Reviewer: The PR there didn't establish the convention. Even before that PR, `callPackage` was used for the vast majority of packages, see [the file before the PR](https://github.com/Mathnerd314/nixpkgs/blob/c545bc80f68a44d0d0e1cc85290d74d1d57c09fa/pkgs/top-level/all-packages.nix).

Here's the actual commit that introduced `callPackage` as a convention: https://github.com/nixos/nixpkgs/commit/d17f0f9cbca38fabb71624f069cd4c0d6feace92

```suggestion
It has [established a convention](https://github.com/nixos/nixpkgs/commit/d17f0f9cbca38fabb71624f069cd4c0d6feace92) of composing parameterised packages with automatic settings through a function named [`callPackage`](https://github.com/NixOS/nixpkgs/commit/fd268b4852d39c18e604c584dd49a611dc795a9b).
```

Author: Uhh thanks for the dig-up!

### What will you learn?

- Using `callPackage` to invoke package recipes that follow Nixpkgs conventions
- Overriding package parameters
- Creating interdependent package sets

### What do you need?

- Familiarity with the [Nix language](reading-nix-language)
- First experience with [packaging existing software](packaging-tutorial)

### How long does it take?

- 45 minutes

## Automatic function calls

Create a new file `hello.nix`, which could be a typical package recipe as found in Nixpkgs:
A function that takes an attribute set, with attributes corresponding to derivations in the top-level package set, and returns a derivation.

```{code-block} nix
:caption: hello.nix
{ writeShellScriptBin }:
writeShellScriptBin "hello" ''
  echo "Hello, world!"
''
```

:::{dropdown} Detailed explanation
`hello.nix` declares a function which takes as argument an attribute set with one element `writeShellScriptBin`.
[`writeShellScriptBin`](https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeShellScriptBin) is a function that happens to exist in Nixpkgs, a [build helper](https://nixos.org/manual/nixpkgs/unstable/#part-builders) that returns a derivation.

> The derivation output in this case is an executable shell script with the contents `echo "hello world"` named `"hello"`.

Reviewer: ```suggestion
The derivation output in this case contains an executable shell script in `$out/bin/hello` that prints "Hello world" when run.
```

The derivation output would be the entire `$out` directory, not just `$out/bin/hello`.

And it doesn't just contain that string, it also contains a shebang before that.

:::

Now create a file `default.nix` with the following contents:

```{code-block} nix
:caption: default.nix
let
  pkgs = import <nixpkgs> { };
in
pkgs.callPackage ./hello.nix { }
```

> Realise the derivation in `default.nix` and run the executable that is produced:

Reviewer: I don't think the terms "realise" and "derivation" work here. The Nix file contains a Nix expression that first needs to be instantiated before it's a derivation.

```suggestion
Build `default.nix` and run the executable that is produced:
```

Author: We haven't really converged on a robust terminology here. In the Nix manual we distinguish between [derivation](https://hydra.nixos.org/build/253249891/download/1/manual/glossary.html#gloss-derivation) (which is rather vaguely defined) and store derivation (the thing that needs to be instantiated). Also "to build" something isn't well defined (although arguably it is the specific act of "running the `builder` executable of a derivation"), while "to realise" is (and is broader than "to build"). 

In short: I'd rather leave that as is, at least for now.

This function takes as arguments the package's dependencies.

## Parameterised builds

Change the `default.nix` to produce an attribute set of derivations, with the attribute `hello` containing the original derivation:

```{code-block} nix
:caption: default.nix
let
  pkgs = import <nixpkgs> { };
in
{
  hello = pkgs.callPackage ./hello.nix { };
}
```

When building the attribute `hello`, by accessing it with the [`-A` / `--attr` option](https://nix.dev/manual/nix/2.19/command-ref/nix-build#opt-attr), the result will be the same as before:

```shell-session
$ nix-build -A hello
$ ./result/bin/hello
Hello, world!
```

Also change `hello.nix` to add an additional parameter `audience` with default value `"world"`:

```{code-block} nix
:caption: hello.nix
{
  writeShellScriptBin,
  audience ? "world",
}:
writeShellScriptBin "hello" ''
  echo "Hello, ${audience}!"
''
```

This also does not change the result.

Things get more interesting when changing `default.nix` to make use of this new argument.
Pass the parameter `audience` in the second argument to `callPackage`:

```{code-block} diff
:caption: default.nix
 let
   pkgs = import <nixpkgs> { };
 in
 {

>    hello = pkgs.callPackage ./hello.nix { };
> +  hello-people = pkgs.callPackage ./hello.nix { audience = "people"; };

Reviewer: This is generally not recommended because any changes to the arguments to `hello` won't influence `hello-people`. It's also conceptually the easiest if each file is only imported from one location. This also leads to the question later when `.override` should be preferred over using another `callPackage`.

I'd say this is best avoided by just passing audience to `hello` instead:

```suggestion
-  hello = pkgs.callPackage ./hello.nix { };
+  hello = pkgs.callPackage ./hello.nix { audience = "people"; };
```

This might warrant moving the introduction of `-A` further down.

 }
```

This attribute is passed on to the argument of the function defined in `hello.nix`:
The same syntax can also be used to explicitly set the automatically discovered arguments, such as `writeShellScriptBin`, but that doesn't make sense here.

Try it out:

```shell-session
$ nix-build -A hello-people
$ ./result/bin/hello
Hello, people!
```

This pattern is used widely in Nixpkgs:
For example, functions which represent Go programs often have a parameter `buildGoModule`.

> It is common to find expressions like `callPackage ./go-program.nix { buildGoModule = buildGo116Module; }` to enforce a certain Go compiler version.

Reviewer: ```suggestion
It is common to find expressions like `callPackage ./go-program.nix { buildGoModule = buildGo116Module; }` to change the default Go compiler version.
```

Not enforced, can be overridden with `.override`.

Nixpkgs is therefore not simply a huge library of pre-configured packages, but a collection of functions – package *recipes* – for customising packages and even entire ecosystems (for example "All Python packages using my custom interpreter") on the fly without duplicating code.

# Overrides

> `callPackage` adds more convenience by adding an attribute to the derivation it returns: the `override` function.
> This means that, as a consequence of parameterising builds with `callPackage`, you can also change the value of these parameters _after the fact_ using the returned derivation's `override` function.

Reviewer: Doesn't flow very well here, the explanation of what the function does should come earlier.

Maybe just

```suggestion
`callPackage` adds more convenience by allowing parameters to be customised _after the fact_ using the returned derivation's `override` function.
```

It's a bit odd to use terms "after" and "customise" when there's no ordering of operations and nothing being changed, but I think it's fine.

Add a third attribute `hello-folks` to `default.nix` and set it to `hello.override` called with a new value for `audience`:

```{code-block} diff
:caption: default.nix
 let
   pkgs = import <nixpkgs> { };
 in
-{
+rec {
   hello = pkgs.callPackage ./hello.nix { };
   hello-people = pkgs.callPackage ./hello.nix { audience = "people"; };
+  hello-folks = hello.override { audience = "folks"; };
 }
```

:::{note}

> The resulting attribute set is now recursive (by the keyword `rec`).

Reviewer: Would be good to link to https://nix.dev/guides/best-practices#recursive-attribute-set-rec here

Author: I've thought about it, but would rather skip it here:
- It's besides the point. Adding more comments will distract from the actual learning objective. The note is merely to remind novice readers what `rec` is about so they can fully grasp the example.
- Referring to a piece that says "`rec` bad" from a place where we very deliberately use `rec` for clarity and small diff seems weird and counterproductive to building trust in the information we provide.
- Thinking about "`rec` bad" made me doubt this should be here on nix.dev to begin with. It seems it's more of a Nixpkgs convention, where it makes sense if you have huge files and the need to separate interface from implementation. But generally, using `rec` judiciously can greatly increase readability in my opinion.

That is, attribute values can refer to names from within the same attribute.
:::

`override` passes `audience` to the original function in `hello.nix` - it *overrides* whatever arguments have been passed in the original `callPackage` that produced the derivation `hello`.
All the other parameters will remain the same.
This is especially useful and can be often found on packages that provide many options to customise a package.

Building `hello-folks` attribute and running the resulting executable will again produce a new version of the script:

```shell-session
$ nix-build -A hello-folks
$ ./result/bin/hello
Hello, folks!
```

> A real-world example is the [`neovim`](https://search.nixos.org/packages?show=neovim) package recipe, which has has overrideable arguments such as `extraLuaPackages`, `extraPythonPackages`, or `withRuby`.

Reviewer: Important to point out is that these parameters are only discoverable by looking at the source file, which people can get to by clicking the "Source" link on search.nixos.org

## Interdependent package sets

You can actually create your own version of `callPackage`!
This comes in handy for package sets where the recipes depend on each other.

:::{note}
The following examples do not show the "called" files, as they are not necessary for understanding the principle.
:::

Consider the following recursive attribute set of derivations:

```{code-block} nix
:caption: default.nix
let
  pkgs = import <nixpkgs> { };
in
rec {
  a = pkgs.callPackage ./a.nix { };
  b = pkgs.callPackage ./b.nix { inherit a; };
  c = pkgs.callPackage ./c.nix { inherit b; };
  d = pkgs.callPackage ./d.nix { };
  e = pkgs.callPackage ./e.nix { inherit c d; };
}
```

:::{note}

> In this example `inherit a;` can be considered equivalent to `a = a;`.

Reviewer: ```suggestion
Here, `inherit a;` is equivalent to `a = a;`.
```

There's some cases where it isn't like in `a: rec { inherit a; }`, but here it is equivalent.

:::

Previously declared derivations are passed as arguments to other derivations through `callPackage`.

In this case you have to remember to manually specify all arguments required by each package in the respective {term}`Nix file` that are not in Nixpkgs.
If `./b.nix` requires an argument `a` but there is no `pkgs.a`, the function call will produce an error.
This can become quite tedious quickly, especially for larger package sets.

> Use `lib.callPackageWith` to create your own `callPackage` based on an attribute set.

Reviewer: Might it not be better to go straight to `lib.makeScope`?

Author: I thought so too until I tried, and it makes things really hairy. I'd rather avoid it, because it would take forever to get right, and the tutorial already provides a lot of value.

The last bit may make your head spin.
This construction is only possible because the Nix language is lazily evaluated.
That is, values are only computed when they are actually needed.
It allows passing `packages` around without having fully defined it.

Each package's dependencies are now implicit at this level (they are still explicit in each of the package files), and `callPackage` resolves them *automagically*.
This relieves you from dealing with them manually, and precludes configuration errors that may only surface late into a lengthy build process.

Of course this small example is still manageable in the original form.
And the implicitly recursive variant can obscure the structure for software developers not familiar with lazy evaluation, making it harder to read for them than it was before.
But this benefit really pays off for large constructions, where it is the amount of code that would obscure the structure, and where manual modifications would become cumbersome and error-prone.

## Summary

> Using `callPackage` does not only follow Nixpkgs conventions, which makes your code easier to follow for experienced Nix users.
> It also gives you some benefits for free:

Reviewer: ```suggestion
Using `callPackage` does not only follow Nixpkgs conventions, which makes your code easier to follow for experienced Nix users, but it also gives you some benefits for free:
```
